### PR TITLE
Feature/improve particle rendering

### PR DESCRIPTION
**Pull request by lyvanettinger**
### Description

image

- added alpha card support for particle rendering
- particles are affected by directional light (very simple implementation)
- particles are affected by directional light's shadows (they don't cast shadows themselves)
- added flags for different rendering settings for particles (can be expanded on)
- added particle stats to drawstats and editor

### Issues

https://bubonic-brotherhood.codecks.io/card/1cx-improve-particle-rendering

### Test criteria

- [ ] Place a physics plane and press 0 (zero) to spawn physics entities with a particle emitter attached on it. By default the particles have the NOSHADOWS flag so they aren't affected by shadows, but are affected by the directional light.
- [ ] Take a look at the lighting calculation in the particle.frag and let me know if you maybe have other suggestions for this (it's the bare minimum, but imo we don't need more)

### CODE REVIEW

File: engine/particles/public/particle_util.hpp [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
 #include <glm/glm.hpp>
 
 // temporary values for testing/progress
-static constexpr uint32_t MAX_EMITTERS = 32;
+static constexpr uint32_t MAX_EMITTERS = 256;
 static constexpr int32_t MAX_PARTICLES = 1024 * 256;
 
+enum class ParticleRenderFlagBits : uint8_t
+{
+    eUnLit = 1 << 0,
```
Comment (MuskettaMan) [inline engine/particles/public/particle_util.hpp:13]: nitpick: Unlit*

File: engine/renderer/private/renderer.cpp [hunk 2]
```diff
@@ -105,7 +105,7 @@ Renderer::Renderer(ApplicationModule& application, Viewport& viewport, const std
     _shadowPipeline = std::make_unique<ShadowPipeline>(_context, *_gBuffers, *_gpuScene);
     _debugPipeline = std::make_unique<DebugPipeline>(_context, *_gBuffers, _uiTarget, *_swapChain);
     _lightingPipeline = std::make_unique<LightingPipeline>(_context, *_gBuffers, _hdrTarget, _brightnessTarget, *_bloomSettings, _ssaoTarget);
-    _particlePipeline = std::make_unique<ParticlePipeline>(_context, _ecs, *_gBuffers, _hdrTarget);
+    _particlePipeline = std::make_unique<ParticlePipeline>(_context, _ecs, *_gBuffers, _hdrTarget, _brightnessTarget, *_bloomSettings);
```
Comment (MuskettaMan) [inline engine/renderer/private/renderer.cpp:108]: shouldnt these dependencies be mentioned in the frame graph?

**Review by MuskettaMan — APPROVED** [review]
Looks good! Tested it on my pc and all works correctly. Left a few comments with questions and nitpicks, but overall good work!

File: shaders/particles/particle.frag [hunk 3]
```diff
@@ -2,16 +2,84 @@
...
+        float cosTheta = max(dot(N, lightDir), 0.0);
+        float bias = CalculateShadowBias(cosTheta, scene.directionalLight.direction.w);
+        DirectionalShadowMap(position, bias, shadow);
+        // TODO: find solution to temporary avoiding of hard shadows
+        shadow += 0.1;
+    }
+
+    if ((flags & UNLIT) != UNLIT)
+    {
+        color *= vec4(scene.directionalLight.color.xyz, 0.0);
```
[the remark this answers is not in the document]
Comment (lyvanettinger) [reply shaders/particles/particle.frag:60]: I can try it out and show how it looks. Personally when I envision what it'd look like it doesn't look like a desirable result but I could be wrong. I assume this is only vertex lighting and no specular? Specular imo would look funky

[the remark this answers is not in the document]
Comment (MuskettaMan) [reply shaders/particles/particle.frag:60]: hmm yea, maybe we should just look at this later, since its pretty related to stylization

**Pull request merged**